source/docs/software/wpilib-tools/robot-simulation/device-sim.rst

```diff
@@ -0,0 +1,113 @@
+Device Simulation
+=================
+
+WPILib provides a way to manage simulation device data in the form of the SimDevice API.
+
+Simulating Core WPILib Device Classes
+-------------------------------------
+
+Core WPILib device classes (i.e ``Encoder``, ``Ultrasonic``, etc.) have simulation classes named ``EncoderSim``, ``UltrasonicSim``, etc. These classes allow interactions with the device data that wouldn't be possible or valid outside of simulation. Using them outside of simulation is undefined behavior - in the best case they will do nothing, worse cases might crash your code!
```

Comment:
> Using them outside of simulation is undefined behavior - in the best case they will do nothing, worse cases might crash your code!

Is this correct? From what I understand, they will not affect your regular robot code. Additionally, we should be clear that "constructing" these classes outside of simulation will not affect anything.

Reply:
Once we start specifying that some stuff will affect and some won't, it becomes a mess. I think that it's better to just say that it is undefined behavior and be done with it.
An example of code crash: some sim functions return `null` out of simulation. That can start a ton of NPEs.

Reply:
The issue is, the Drivetrain Simulation tutorial says it won't interfere, and that conflicts with this.

Reply:
![image](https://user-images.githubusercontent.com/10674555/111078251-e4187380-84ca-11eb-9c24-dea8540d5687.png)

Reply:
I think individuals doing things in simulation will properly understand enough to get the gist of things.